### PR TITLE
Address empty cluster slots list when discovering slots

### PR DESCRIPTION
**Pull request by jgaoIXL**
This is a bug that I encountered where `renewClusterSlots` was called when one node was removed from the cluster. The removed node happened to get selected to do the `discoverClusterSlots`. Because the removed node return no cluster slots, we have an empty slots cache, and there is no recovery possible from there since the nodes and slots map are all empty. 

I think it makes sense to return if slotsInfo is size 0 because under no normal circumstances, the cluster should have 0 cluster info provided, it can only mean one thing, which is you are pointing to a node that is not in the cluster.

### CODE REVIEW

**Review by zeekling — DISMISSED** [review]
It seems reasonable.

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Let's use `isEmpty()` method.

**Comment by jgaoIXL** [comment]
> Let's use `isEmpty()` method.

Fixed, also `discoverClusterNodesAndSlots` should probably throw an exception when this happens.

**Comment by yangbodong22011** [comment]
@jgaoIXL Jedis assumes that the `initial node` of the cluster is `constant` (that is, it always belongs to this cluster)，whether it is vip, domain, or physical ip.

But I agree to merge this PR, as a defensive protection.

File: src/main/java/redis/clients/jedis/JedisClusterInfoCache.java [hunk 1]
```diff
@@ -139,6 +143,9 @@ public void renewClusterSlots(Connection jedis) {
 
   private void discoverClusterSlots(Connection jedis) {
     List<Object> slotsInfo = executeClusterSlots(jedis);
+    if (slotsInfo.isEmpty()) {
+      return;
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/JedisClusterInfoCache.java:147]:
```suggestion
      throw new JedisClusterException("Cluster slots list is empty");
```

Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/JedisClusterInfoCache.java:147]: throw JedisClusterException, catch and ignore it, and try next nodes.

Comment (jgaoIXL) [reply src/main/java/redis/clients/jedis/JedisClusterInfoCache.java]: Fixed

**Comment by yangbodong22011** [comment]
[here](https://github.com/redis/jedis/pull/3195/files#diff-4a02ef8d8a92f62e06be741dac2e739e37916ea33c7f0cb5b82ab040361adbcbR106) need catch `JedisClusterException`.

File: src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java [hunk 2]
```diff
@@ -41,7 +42,7 @@ private void initializeSlotsCache(Set<HostAndPort> startNodes, JedisClientConfig
       try (Connection jedis = new Connection(hostAndPort, clientConfig)) {
         cache.discoverClusterNodesAndSlots(jedis);
         return;
-      } catch (JedisConnectionException e) {
+      } catch (JedisConnectionException | JedisClusterException e) {
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:45]:
```suggestion
      } catch (JedisException e) {
```

File: src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java [hunk 3]
```diff
@@ -14,6 +14,7 @@
 import redis.clients.jedis.Connection;
 import redis.clients.jedis.ConnectionPool;
 import redis.clients.jedis.JedisClusterInfoCache;
+import redis.clients.jedis.exceptions.JedisClusterException;
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:17]:
```suggestion
```

Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:17]: Consistent with the rest.

File: src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java (hunk 2 above)
Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:45]: Consistent with the rest.

File: src/main/java/redis/clients/jedis/JedisClusterInfoCache.java [hunk 4]
```diff
@@ -58,6 +59,9 @@ public JedisClusterInfoCache(final JedisClientConfig clientConfig,
 
   public void discoverClusterNodesAndSlots(Connection jedis) {
     List<Object> slotsInfo = executeClusterSlots(jedis);
+    if (slotsInfo.isEmpty()) {
+      throw new JedisClusterException("Cluster slots list is empty");
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisClusterInfoCache.java:63]:
```suggestion
      throw new JedisClusterOperationException("Cluster slots list is empty.");
```

File: src/main/java/redis/clients/jedis/JedisClusterInfoCache.java [hunk 5]
```diff
@@ -14,6 +14,7 @@
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
 
+import redis.clients.jedis.exceptions.JedisClusterException;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisClusterInfoCache.java:17]:
```suggestion
import redis.clients.jedis.exceptions.JedisClusterOperationException;
```

File: src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java (hunk 3 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:17]:
```suggestion
```

File: src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java (hunk 2 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java:45]:
```suggestion
      } catch (JedisConnectionException | JedisClusterOperationException e) {
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]
The exception should be `JedisClusterOperationException`. WIP #3204

**Review by sazzad16 — APPROVED** [review]

**Review by yangbodong22011 — APPROVED** [review]

**Comment by jgaoIXL** [comment]
Thank you. When are we supposed to see a stable release with this fix? @yangbodong22011

**Comment by sazzad16** [comment]
@jgaoIXL We don't have an exact date yet but this change is available in 4.4.0 snapshot.

**Pull request merged**